File: utils/model_utils.py
import numpy as np
import libpysal.weights as weights
import pysal.explore as esda
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

def separate_features(df):
    # Drop R columns
    df = df.drop(columns = [col for col in df.columns if "fold_id_r" in col])
    
    # Dependent variables
    labels = df.pop('very_good_health')

    # Outer CV folds
    outer_fold_ids = df["outer_loop_fold_id_python"]
    outer_splits = np.sort(outer_fold_ids.unique().astype(int))

    # Inner CV folds
    inner_fold_ids = df[[col for col in df.columns if "inner_loop" in col]]
    inner_splits = np.sort(inner_fold_ids.stack().unique().astype(int))

    # Independent variables
    features = df.drop(columns = [col for col in df.columns if "fold_id" in col])
    features["x_coord"] = features["geometry"].centroid.x
    features["y_coord"] = features["geometry"].centroid.y
    features = features.drop(columns = ["geometry"])

    return labels, outer_fold_ids, outer_splits, inner_fold_ids, inner_splits, features


def split_data(current_split, fold_ids, features, labels, is_outer = False, inner_fold_ids = None):
    is_in_validation_set = fold_ids == current_split
    is_in_training_set = ~is_in_validation_set
    train_features = features.loc[is_in_training_set]
    train_labels = labels.loc[is_in_training_set]
    val_features = features.loc[is_in_validation_set]
    val_labels = labels.loc[is_in_validation_set]
    current_inner_fold_ids = None
    if is_outer:
        current_inner_fold_ids = inner_fold_ids.loc[is_in_training_set]
    return train_features, train_labels, val_features, val_labels, current_inner_fold_ids

def get_ann_random_hyperparameters(features):
    no_of_layers = np.random.randint(1, 5)
    no_of_nodes = []
    for i in range(0, no_of_layers):
        no_of_nodes.append(np.random.randint((len(features.columns) / 2), (len(features.columns) * 2)))
    learning_rate = np.random.uniform(0.0001, 0.1)
    return no_of_layers, no_of_nodes, learning_rate

def get_gwr_random_hyperparameters():
    kernel = np.random.choice(["bisquare", "Gaussian", "exponential"])
    criterion = np.random.choice(["AICc", "AIC", "BIC", "CV"])
    return kernel, criterion    

def get_evaluation_metrics(val_features, val_labels, predictions):
    mae = mean_absolute_error(val_labels, predictions)
    mse = mean_squared_error(val_labels, predictions)
    r2 = r2_score(val_labels, predictions)
    return mae, mse, r2

def get_avg_scores(cv_results):
    mae_scores = []
    mse_scores = []
    r2_scores = []

    for result in cv_results:
        mae_scores.append(result["mae"])
        mse_scores.append(result["mse"])
        r2_scores.append(result["r2"])

    avg_mae = np.mean(mae_scores)
    avg_mse = np.mean(mse_scores)
    avg_r2 = np.mean(r2_scores)

    return avg_mae, avg_mse, avg_r2

def get_optimal_hyperparameters(hp_combinations, cv_results):
    hp_combination_scores = []
    for i in range(len(hp_combinations)):
        current_hp_combination_results = [result for result in cv_results if result["hp_combination"] == i]
        mae, mse, r2 = get_avg_scores(current_hp_combination_results)
        hp_combination_scores.append(mse)
    optimal_combination = np.argmin(hp_combination_scores)
    optimal_hps = hp_combinations[optimal_combination]
    return optimal_hps

File: utils/model_utils.R
# Function - Create spatial weight matrix
library(spdep)
library(sf)
library(sfdep)
library(Metrics)

# Separate cross-validation fold ids from df and drop unneeded columns
prepare_data <- function(df) {
  drop_cols <- names(df)[grepl("python", names(df))]
  df <- df[, !names(df) %in% drop_cols]

  # Extract fold ids
  outer_fold_ids <- df$outer_loop_fold_id_r
  outer_splits <- sort(unique(outer_fold_ids))
  inner_fold_ids <- df[, grepl("inner_loop", names(df)), drop = FALSE]
  inner_splits <- c(1, 2, 3, 4, 5)

  # Drop unneeded cols
  drop_cols <- names(df)[grepl("fold_id", names(df))]
  df <- df[, !names(df) %in% drop_cols]

  prepared_data <- list(
    outer_fold_ids = outer_fold_ids,
    outer_splits = outer_splits,
    inner_fold_ids = inner_fold_ids,
    inner_splits = inner_splits,
    df = df
  )

  # Return separated data
  prepared_data
}


# Split df into training and validation sets
split_data <- function(df, current_split, fold_ids, is_outer = FALSE, inner_fold_ids = NULL) {
  is_in_validation_set <- fold_ids == current_split
  is_in_training_set <- !is_in_validation_set
  train_df <- df[is_in_training_set, ]
  val_df <- df[is_in_validation_set, ]

  # If splitting data for outer cross-validation loop, get fold ids for inner cross-validation loop
  current_inner_fold_ids <- NULL
  if (is_outer) {
    current_inner_fold_ids <- inner_fold_ids[is_in_training_set, , drop = FALSE] # Removes rows not in current training set
    rownames(current_inner_fold_ids) <- NULL # Reset row numbers
  }
  rownames(train_df) <- NULL
  rownames(val_df) <- NULL

  split_data <- list(
    train_df = train_df,
    val_df = val_df,
    current_inner_fold_ids = current_inner_fold_ids
  )

  # Return split data
  split_data
}


create_spatial_weights <- function(df, hps) {
  weighting_method <- hps$weighting_method
  coords <- st_centroid(st_geometry(df))
  w <- NULL

  if (weighting_method == "distance") {
    print(weighting_method)
    print(hps$max_distance)
    dnn <- dnearneigh(coords, d1 = 200, d2 = hps$max_distance)
    w <- nb2listw(dnn, zero.policy = TRUE)
  } else if (weighting_method == "knn") {
    print(weighting_method)
    print(hps$k)
    knn <- knearneigh(coords, k = hps$k)
    w <- nb2listw(knn2nb(knn), zero.policy = TRUE)
  } else {
    print(weighting_method)
    print(hps$is_queen)
    nb <- st_contiguity(df, hps$is_queen)
    w <- nb2listw(nb, zero.policy = TRUE)
  }

  # Return weights
  w
}

get_evaluation_metrics <- function(labels, predictions) {
  mae <- mae(labels, predictions)
  mse <- mse(labels, predictions)
  sum_of_squares_residual <- sum((labels - predictions)^2)
  sum_of_squares_total <- sum((labels - mean(labels))^2)
  r2 <- 1 - (sum_of_squares_residual / sum_of_squares_total)
  list(mae = mae, mse = mse, r2 = r2)
}

# Get average accuracy metrics from set of cross-validation results
get_avg_scores <- function(cv_results) {
  mae_scores <- c()
  mse_scores <- c()
  r2_scores <- c()

  for (result in cv_results) {
    mae_scores <- c(mae_scores, result$mae)
    mse_scores <- c(mse_scores, result$mse)
    r2_scores <- c(r2_scores, result$r2)
  }

  mae <- mean(mae_scores)
  mse <- mean(mse_scores)
  r2 <- mean(r2_scores)

  avgs <- list(mae = mae, mse = mse, r2 = r2)
  avgs
}

# Get best performing hyperparameters from set of cross-validation results
get_optimal_hps <- function(hp_combinations, cv_results) {
  hp_combination_scores <- c()

  for (hp_combination in seq_along(hp_combinations)) {
    current_hp_combination_results <- cv_results[sapply(cv_results, function(result) result$hp_combination == hp_combination)]
    avg_scores <- get_avg_scores(current_hp_combination_results)
    mae <- avg_scores$mae
    mse <- avg_scores$mse
    r2 <- avg_scores$r2
    hp_combination_scores <- c(hp_combination_scores, mae)
  }

  optimal_combination <- which.min(hp_combination_scores)
  optimal_hps <- hp_combinations[[optimal_combination]]
  optimal_hps
}
